src/dbus/dbus_resources.hpp

```diff
@@ -29,61 +29,21 @@
 #ifndef OTBR_DBUS_DBUS_RESOURCES_HPP_
 #define OTBR_DBUS_DBUS_RESOURCES_HPP_
 
+#include <memory>
 #include <utility>
 
 #include <dbus/dbus.h>
 
 namespace otbr {
 namespace dbus {
 
-template <typename T, T *(*refFunc)(T *), void (*unrefFunc)(T *)> class SharedDBusResource final
+template <typename kType, void (*kUnref)(kType *)> struct DBusDeleter
 {
-public:
-    explicit SharedDBusResource(T *aResource)
-        : mResource(aResource)
-    {
-        refFunc(aResource);
-    }
-
-    SharedDBusResource(const SharedDBusResource &aOther) { CopyFrom(aOther); }
-
-    SharedDBusResource &operator=(const SharedDBusResource &aOther) { return CopyFrom(aOther); }
-
-    SharedDBusResource(SharedDBusResource &&aOther) { Reset(std::move(aOther)); }
-
-    SharedDBusResource &operator=(SharedDBusResource &&aOther) { return Reset(std::move(aOther)); }
-
-    T *operator->(void) { return mResource; }
-    T *GetRaw(void) { return mResource; }
-
-    ~SharedDBusResource(void)
-    {
-        unrefFunc(mResource);
-        mResource = nullptr;
-    }
-
-private:
-    SharedDBusResource &CopyFrom(const SharedDBusResource &aOther)
-    {
-        this->mResource = aOther.mResource;
-        refFunc(this->mResource);
-
-        return *this;
-    }
-
-    SharedDBusResource &Reset(SharedDBusResource &&aOther)
-    {
-        this->mResource  = aOther.mResource;
-        aOther.mResource = nullptr;
-
-        return *this;
-    }
-
-    T *mResource;
+    void operator()(kType *aPointer) { kUnref(aPointer); }
 };
 
-using SharedDBusConnection = SharedDBusResource<DBusConnection, dbus_connection_ref, dbus_connection_unref>;
-using SharedDBusMessage    = SharedDBusResource<DBusMessage, dbus_message_ref, dbus_message_unref>;
+using UniqueDBusMessage    = std::unique_ptr<DBusMessage, DBusDeleter<DBusMessage, &dbus_message_unref>>;
+using UniqueDBusConnection = std::unique_ptr<DBusConnection, DBusDeleter<DBusConnection, &dbus_connection_unref>>;
```

Comment:
Should eliminate `dbus_message_unref` anywhere else and always use `UniqueDBusMessage`? The same for `DBusConnection`?

Reply:
The only occurrence of `dbus_message_unref` is in dbus_request.hpp which we've agreed to use raw pointer.

Reply:
what about `ncp_wpantund.cpp` and `dbus_base.cpp` ?

Reply:
We won't touch those code not under dbus/ folder in this PR.